0010-endpoint-metadata-extension/README.md

```diff
@@ -0,0 +1,687 @@
+---
+name: Endpoint Metadata Extension
+shortname: 10/EPMDX
+status: draft
+editor: Volodymyr Tkhir <vtkhir@cybervisiontech.com>
+---
+
+- [Introduction](#introduction)
+- [Requirements and constraints](#requirements-and-constraints)
+  - [Metadata keys](#metadata-keys)
+  - [Metadata values](#metadata-values)
+  - [Supported operations](#supported-operations)
+- [Design](#design)
+  - [Full metadata update](#full-metadata-update)
+    - [Full metadata update request](#full-metadata-update-request)
+    - [Full metadata update response](#full-metadata-update-response)
+  - [Partial metadata update](#partial-metadata-update)
+    - [Partial metadata update request](#partial-metadata-update-request)
+    - [Partial metadata update response](#partial-metadata-update-response)
+  - [Get metadata](#get-metadata)
+    - [Get metadata request](#get-metadata-request)
+    - [Get metadata response](#get-metadata-response)
+  - [Get metadata keys](#get-metadata-keys)
+    - [Get metadata keys request](#get-metadata-keys-request)
+    - [Get metadata keys response](#get-metadata-keys-response)
+  - [Delete metadata key](#delete-metadata-key)
+    - [Delete metadata key request](#delete-metadata-key-request)
+    - [Delete metadata key response](#delete-metadata-key-response)
+
+
+## Introduction
+
+The Endpoint Metadata Extension (EPMDX) protocol is an endpoint-aware [Kaa Protocol](/0001-kaa-protocol/README.md) extension.
+
+It is intended to manage endpoint metadata.
+
+## Requirements and constraints
+
+### Metadata keys
+EPMDX protocol supports only string type of metadata keys.
+All UTF-8 characters are supported in metadata key names.
+Metadata key must contain at least one non-whitespace character.
```

Comment:
We have already discussed that with @vtkhir and we agreed to support empty/whitespace-only keys. The main reason is "why not?"

I mean, there is no reason to reject them as this is a benefit for neither implementation nor usage. (Uniqueness is not an argument here, because `""` could be unique.)

Furthermore, if you don't see a use case now, that is not the reason to forbid the usage as a user may come with a use case later and then they are doomed to implement hacks and workarounds.

Finally, I think this complicates implementations as there is an additional check now (which includes decoding Unicode strings?). Also, "whitespace" is not defined. Does it include spaces? tabs? newlines? carriage return? no-break space? All Unicode whitespaces?

I'd rather treat keys as opaque strings and wouldn't make any assumptions (except for uniqueness, of course).